src/base/bittorrent/tracker.cpp

```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
 
 using namespace BitTorrent;
 
-// Peer
-bool Peer::operator!=(const Peer &other) const
+using std::string;
+using boost::asio::ip::address;
+
+namespace
 {
-    return uid() != other.uid();
+    bool isNonRoutable(const boost::asio::ip::address_v4& ip)
+    {
+        if (ip.is_loopback() || ip.is_unspecified() || ip.is_multicast())
+            return true;
+
+        quint64 addr = ip.to_ulong();
+
+        return ((addr >> 24) == 0x0A)    // 10.0.0.0/8      Private
+            || ((addr >> 20) == 0xAC1)   // 172.16.0.0/12   Private
+            || ((addr >> 16) == 0xC0A8)  // 192.168.0.0/16  Private
+            || ((addr >> 16) == 0xA9FE); // 169.254.0.0/16  Link-local
+    }
+
+    boost::asio::ip::address_v4 toAddress(uchar a, uchar b, uchar c, uchar d)
+    {
+        return boost::asio::ip::address_v4((a << 24) + (b << 16) + (c << 8) + d);
+    }
+
+    bool isNonRoutable(const boost::asio::ip::address& ip)
+    {
+        if (ip.is_loopback() || ip.is_unspecified() || ip.is_multicast())
+            return true;
+
+        if (ip.is_v4())
+            return isNonRoutable(ip.to_v4());
+
+        if (ip.is_v6()) {
+            boost::asio::ip::address_v6 ipv6 = ip.to_v6();
+            auto addr = ipv6.to_bytes();
+            return ((addr[0] >> 1) == (0xFC >> 1))  // FC00::/7   Unique local
+                || ipv6.is_link_local()     // FE80::/10  Link-local
+                // ::ffff:0:0/96 ipv4-mapped of non-routable address
+                || (ipv6.is_v4_mapped() && isNonRoutable(ipv6.to_v4()))
+                // 2002::/16 6to4 ipv6 address
+                || (addr[0] == 0x20 && addr[1] == 0x02
+                    && isNonRoutable(toAddress(addr[2], addr[3], addr[4], addr[5])))
+                // 64:ff9b::/96
+                || (addr[0] == 0 && addr[1] == 0x64 && addr[2] == 0xFF && addr[3] == 0x9B
+                    && std::all_of(&addr[4], &addr[12], [](char a) { return a == 0; })
+                    && isNonRoutable(toAddress(addr[12], addr[13], addr[14], addr[15])));
+        }
+
+        return true;
+    }
+
+    // percent-escapes given string
+    std::string escape(const QByteArray& str)
+    {
+        // like libtorrent percent-encoding
+        QByteArray encoded = str.toPercentEncoding("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~!()*");
```

Comment:
Although I haven't tried it myself, Qt5 documentation example for this function shows that characters from the `exclude` parameter are used as addition to the standard set of [a-zA-Z0-9] and "-" / "." / "_" / "~". Thus, I believe, the call can be simplified:
```C++
QByteArray encoded = str.toPercentEncoding("!()*");
```